### PR TITLE
Add signature inversion functions, test, and docs

### PR DESCRIPTION
**Pull request by afermanian**
Implement the insertion algorithm for signature inversion.

### CODE REVIEW

File: docs/pages/examples/inversion.rst [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+.. code-block:: python
+
+    import numpy as np
+    import torch
+    import signatory
+
+    # Create a path consisting in a half circle
+    time = np.linspace(0, 1, 10)
+    path = torch.from_numpy(np.stack([np.cos(np.pi * time), np.sin(np.pi * time)])).T.unsqueeze(0)
```
Comment (patrick-kidger) [inline docs/pages/examples/inversion.rst:16]: Doing this in just PyTorch would be preferred: `torch.linspace` etc.

File: docs/pages/reference/signatures-inversion.rst [hunk 2]
```diff
@@ -0,0 +1,11 @@
+.. _reference-signatures-inversion:
+
+Signature inversion
+###################
+
+.. currentmodule:: signatory
+
+See [...] for a description of the insertion algorithm.
```
Comment (patrick-kidger) [inline docs/pages/reference/signatures-inversion.rst:8]: This `...` looks like it needs filling in?

File: src/signatory/__init__.py [hunk 3]
```diff
@@ -50,6 +50,7 @@
                                extract_signature_term,
                                signature_combine,
                                multi_signature_combine)
+from .signature_inversion_module import (invert_signature)
```
Comment (patrick-kidger) [inline src/signatory/__init__.py:53]: Nit: these brackets are unnecessary.

File: src/signatory/signature_inversion_module.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+def invert_signature(signature: torch.Tensor, depth: int, channels: int,
+                     initial_position: Optional[torch.Tensor] = None) -> torch.Tensor:
+    r"""Invert the signature with the insertion algorithm: reconstruct a stream of data given its signature. Given that
+    the signature is invariant by translation, the initial position of the stream is not recovered.
+
+    The input :attr:`signature` is the signature transform of depth :attr:`depth` of a batch of paths: it should be a
+    result from the :func:`signatory.signature` function. The output is a tensor of shape :math:`(N, L, C)`,
+    where :math:`N` is the batch size, :math:`L` is the length of the reconstructed stream of data, with
+    :math:`L = depth + 1`, and :math:`C` denotes the number of channels.
+
```
Comment (patrick-kidger) [inline src/signatory/signature_inversion_module.py:80]: Can you add a reference to the paper in the docstring?

File: test/test_signature_inversion.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+def test_inverted_path_shape():
+    """Tests that the inverted path is of the right shape"""
+    for batch_size in (1, 2, 5):
+        for input_stream in (2, 3, 10):
+            for input_channels in (1, 2, 6):
+                path = torch.rand((batch_size, input_stream, input_channels))
+                for depth in (2, 4, 6):
+                    signature = signatory.signature(path, depth)
+                    inverted_path = signatory.invert_signature(signature, depth, input_channels, initial_position=path[:, 0, :])
+                    assert inverted_path.shape == (batch_size, depth + 1, input_channels)
```
Comment (patrick-kidger) [inline test/test_signature_inversion.py:21]: It would be nice to have some tests validating that the inversion produces a result close to the original path. Possibly by taking a short path length and high signature depth, to be able to do a good inversion?

If need be you can control the randomness by passing a seed to `torc.rand`, so that you can pick a seed that produces good results, rather than one that fails the inversion check every now and again.

**Review by patrick-kidger — COMMENTED** [review]
Various minor comments aside, this looks good to me!

**Comment by patrick-kidger** [comment]
Can you bump the version number + changelog as well please?

File: docs/pages/examples/inversion.rst (hunk 1 above)
Comment (afermanian) [reply docs/pages/examples/inversion.rst:16]: Done !

File: docs/pages/reference/signatures-inversion.rst (hunk 2 above)
Comment (afermanian) [reply docs/pages/reference/signatures-inversion.rst:8]: Yes this is a typo. I have not yet put the paper on arXiv, I will add the reference when I do. Would that be ok to you?

File: src/signatory/__init__.py (hunk 3 above)
Comment (afermanian) [reply src/signatory/__init__.py:53]: Done

File: src/signatory/signature_inversion_module.py (hunk 4 above)
Comment (afermanian) [reply src/signatory/signature_inversion_module.py:80]: See above, I will add the reference when I put the paper on arXiv.

File: test/test_signature_inversion.py (hunk 5 above)
Comment (afermanian) [reply test/test_signature_inversion.py:21]: Yes I agree. I have added a test with the half circle example. Note that you have to set the tolerance very high (1e-01) to pass the test but I guess it is better than nothing.

**Comment by afermanian** [comment]
> Can you bump the version number + changelog as well please?

Do you mean I should do a new version ? (1.2.5)

**Comment by patrick-kidger** [comment]
Yep, that's what I mean. Increment the version counter here:

https://github.com/patrick-kidger/signatory/blob/7712a9c032473d2c2ee5b16a0fc0c170c7d87766/src/signatory/__init__.py#L59

And add a brief description of your changes to the changelog here:

https://github.com/patrick-kidger/signatory/blob/master/CHANGELOG.txt

File: test/test_signature_inversion.py [hunk 6]
```diff
@@ -34,3 +33,19 @@ def test_initial_position_zero():
...
+    the original one."""
+    input_stream = 10
+    time = torch.linspace(0, 1, input_stream)
+    path = torch.stack([torch.cos(3.14 * time), torch.sin(3.14 * time)]).T.unsqueeze(0)
+    input_channels = 2
+    depth = input_stream - 1
+    signature = signatory.signature(path, depth)
+    inverted_path = signatory.invert_signature(signature, depth, input_channels, initial_position=path[:, 0, :])
+
+    assert torch.all(torch.isclose(path[:, :, :], inverted_path[:, :, :], atol=1e-01))
```
Comment (patrick-kidger) [inline test/test_signature_inversion.py:50]: Nit: `torch.allclose` instead of `torch.all(torch.isclose(...))`

**Comment by patrick-kidger** [comment]
It looks like the readme check is failing. Which is a little surprising, I didn't think you'd touched anything that affected that. In any case run `python command.py readme` and it should re-create the readme automatically from the documentation, and the check will pass.

Very happy to accept a subsequent PR adding references to the arXiv paper once it's up.

**Comment by patrick-kidger** [comment]
@afermanian And merged! Thanks for contributing.
